kedro/extras/datasets/video/video_dataset.py

```diff
@@ -0,0 +1,280 @@
+    def fourcc(self) -> str:
+        fourcc = self._cap.get(cv2.CAP_PROP_FOURCC)
+        return int(fourcc).to_bytes(4, "little").decode("ascii")
+
+    @property
+    def fps(self) -> float:
+        return self._cap.get(cv2.CAP_PROP_FPS)
+
+    @property
+    def size(self) -> Tuple[int, int]:
+        width = int(self._cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+        height = int(self._cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        return width, height
+
+    def __getitem__(self, index: Union[int, slice]):
+        if isinstance(index, slice):
+            return SlicedVideo(self, index)
+
+        if index < 0:
+            index += len(self)
+        if index >= len(self):
+            raise IndexError()
+
+        if index != self._index:
+            self._cap.set(cv2.CAP_PROP_POS_FRAMES, index)
+        self._index = index + 1  # Next frame to decode after this
+        ret, frame_bgr = self._cap.read()
+        if not ret:
+            raise IndexError()
+
+        height, width = frame_bgr.shape[:2]
+        return PIL.Image.frombuffer(  # Convert to PIL image with RGB instead of BGR
+            "RGB", (width, height), frame_bgr, "raw", "BGR", 0, 0
+        )
+
+    def _get_length(self) -> int:
+        # OpenCV's frame count might be an approximation depending on what
+        # headers are available in the video file
+        length = int(round(self._cap.get(cv2.CAP_PROP_FRAME_COUNT)))
+        if length >= 0:
+            return length
+
+        # On some video files it fails, we can count the frames but it will be
+        # very slow... It also seems like that opencv has issues to seek in those
+        # videos...
+        raise ValueError("Failed to load video since number of frames can't be infered")
+
+
+class SequenceVideo(AbstractVideo):
+    """A video object read from an indexable sequence of frames"""
+
+    def __init__(
+        self, frames: Sequence[PIL.Image.Image], fps: float, fourcc: str = "mp4v"
+    ) -> None:
+        self._n_frames = len(frames)
+        self._frames = frames
+        self._fourcc = fourcc
+        self._size = frames[0].size
+        self._fps = fps
+
+    @property
+    def fourcc(self) -> str:
+        return self._fourcc
+
+    @property
+    def fps(self) -> float:
+        return self._fps
+
+    @property
+    def size(self) -> Tuple[int, int]:
+        return self._size
+
+    def __getitem__(self, index: Union[int, slice]):
+        if isinstance(index, slice):
+            return SlicedVideo(self, index)
+        return self._frames[index]
+
+
+class GeneratorVideo(AbstractVideo):
+    """A video object with frames yielded by a generator"""
+
+    def __init__(
+        self,
+        frames: Generator[PIL.Image.Image, None, None],
+        length,
+        fps: float,
+        fourcc: str = "mp4v",
+    ) -> None:
+        self._n_frames = length
+        self._gen = more_itertools.peekable(frames)
+        self._fourcc = fourcc
+        self._size = self._gen.peek().size
```

Comment:
@daniel-falk Just re-iterating this comment as this could be a bug potentially.

Reply:
In your example `peek` will be chained to include the first element even after peeking. In my code this is `self._gen`, this is what is used later. `frames` will be missing the first element after peeking, but that is not used anymore. Your example should instead be:
```
In [14]: def gen():
    ...:     for i in range(5):
    ...:         yield i
    ...: 
In [15]: 
In [15]: g = gen()
In [16]: p = more_itertools.peekable(g)
In [17]: p.peek()
Out[17]: 0
In [18]: list(p)
Out[18]: [0, 1, 2, 3, 4]
```

I changed the implementation to use regular itertools instead of more_itertools.